src/shared/image_viewer.rs

```diff
@@ -0,0 +1,1134 @@
+    event::TouchUpdateEvent,
+    image_cache::{ImageBuffer, ImageError},
+    rotated_image::RotatedImageWidgetExt,
+    *,
+};
+
+use crate::shared::{avatar::{AvatarRef, AvatarWidgetRefExt}, timestamp::TimestampWidgetRefExt};
+
+/// Loads the given image `data` into an `ImageBuffer` as either a PNG or JPEG, using the `imghdr` library to determine which format it is.
+///
+/// Returns an error if either load fails or if the image format is unknown.
+///
+pub fn get_png_or_jpg_image_buffer(data: Vec<u8>) -> Result<ImageBuffer, ImageError> {
+    match imghdr::from_bytes(&data) {
+        Some(imghdr::Type::Png) => {
+            ImageBuffer::from_png(&data)
+        },
+        Some(imghdr::Type::Jpeg) => {
+            ImageBuffer::from_jpg(&data)
+        },
+        Some(_unsupported) => {
+            // Attempt to load it as a PNG or JPEG anyway, since imghdr isn't perfect.
+            Err(ImageError::UnsupportedFormat)
+        }
+        None => {
+            // Attempt to load it as a PNG or JPEG anyway, since imghdr isn't perfect.
+            Err(ImageError::UnsupportedFormat)
+        }
+    }
+}
+
+/// Duration for rotation animations in seconds.
+/// This value should be consistent with the duration value in set in the animator.
+const ROTATION_ANIMATION_DURATION: f64 = 1.0;
+
+/// Configuration for zoom and pan settings in the image viewer
+#[derive(Clone, Debug)]
+pub struct Config {
+    /// Minimum zoom level (default: 0.5)
+    pub min_zoom: f32,
+    /// Maximum zoom level (default: 4.0)
+    pub max_zoom: f32,
+    /// Zoom scale factor for zoom in/out operations (default: 1.2)
+    pub zoom_scale_factor: f32,
+    /// Pan sensitivity multiplier for drag operations (default: 2.0)
+    pub pan_sensitivity: f64,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            min_zoom: 0.5,
+            max_zoom: 4.0,
+            zoom_scale_factor: 1.2,
+            pan_sensitivity: 2.0,
+        }
+    }
+}
+
+/// Error types for image loading operations
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ImageViewerError {
+    /// Image not found
+    NotFound,
+    /// Access denied
+    Unauthorized,
+    /// Connection failed
+    ConnectionFailed,
+    /// Bad or corrupted data
+    BadData,
+    /// Server error
+    ServerError,
+    /// Unsupported format
+    UnsupportedFormat,
+    /// Unknown error
+    Unknown,
+    /// Image loading timed out
+    Timeout,
+}
+
+/// Returns a human-readable string describing the given ImageViewerError.
+///
+/// This can be used to update the label text of an error display.
+///
+/// The error type is matched against a string which describes the error in a way that is visible to the user.
+///
+/// The strings returned by this function will be appropriate for display in a label or similar widget.
+pub fn image_viewer_error_to_string(error: &ImageViewerError) -> &str {
+    match error {
+        ImageViewerError::NotFound => "Full image is not found",
```

Comment:
nit

```suggestion
        ImageViewerError::NotFound => "Full image was not found",
```

Also, what does this actually mean? the source URI was bad, or the image itself has been deleted, or something else? (i'm just curious, the error message itself is probably fine)